main.tf

```diff
@@ -34,23 +34,64 @@ locals {
       }]) } : {},
   )]
 
-  #Remove all options that are null
-  workflows = merge(
-    var.use_predefined_workflows ? yamldecode(file("${path.module}/config/workflows.yaml")).workflows : {},
-    {
-      for workflow_name, workflow in var.workflows : workflow_name => {
-        for k, v in workflow : k => v if v != null
-      }
+  workflows_helper_options = ["asdf", "checkov", "pull_gitlab_variables", "check_gitlab_approvals", "template"]
+
+  pre_workflows = {
+    for workflow_name, workflow in var.workflows : workflow_name => {
+      plan = merge(
+        local.null_workflow.plan,
+        can(workflow.template) ? local.workflow_templates[workflow["template"]].plan : null,
+      workflow.plan)
+      apply = merge(
+        local.null_workflow.apply,
+        can(workflow.template) ? local.workflow_templates[workflow["template"]].apply : null,
+      workflow.apply),
+      import = merge(
+        local.null_workflow.import,
+        can(workflow.template) ? local.workflow_templates[workflow["template"]].import : null,
+      workflow.import)
+      state_rm = merge(
+        local.null_workflow.state_rm,
+        can(workflow.template) ? local.workflow_templates[workflow["template"]].state_rm : null,
+      workflow.state_rm),
+      pull_gitlab_variables  = workflow.pull_gitlab_variables
+      asdf                   = workflow.asdf
+      checkov                = workflow.checkov
+      check_gitlab_approvals = workflow.check_gitlab_approvals
     }
-  )
+  }
+
+  workflows = {
+    for workflow_name, workflow in local.pre_workflows : workflow_name => {
+      for stage_name, stage in workflow : stage_name => { steps : concat(
+        workflow.asdf.enabled && stage_name == "plan" ? [{ run = "asdf install" }] : [],
+        workflow.pull_gitlab_variables.enabled ? [{ multienv = "pull_gitlab_variables.sh" }] : [],
+        workflow.check_gitlab_approvals.enabled && stage_name == "apply" ? [{ run = "check_gitlab_approvals.sh" }] : [],
+        flatten([
+          for step in stage.steps : [
+            for name, object in step :
+            jsondecode(
+              name == "atlantis_step" ?
+              (object.extra_args != null ? jsonencode({ (object.command) : { extra_args : object.extra_args } }) : jsonencode(object.command)) :
+              jsonencode({ (name) : object })
+            )
+            if object != null
+        ]]),
+        [for e in ["show", { run = format("checkov -f $SHOWFILE -o github_failed_only %s", workflow.checkov.soft_fail ? "--soft-fail" : "") }] : e
```

Comment:
for now I decided to do soft-fail flag inline but in the future it'd be better to maybe make submodule out of that...?